lib/flv/transmuxer.js

```diff
@@ -0,0 +1,563 @@
+    }
+
+    while (nalUnits.length) {
+      currentNal = nalUnits.shift();
+
+    // record the track config
+    if (currentNal.nalUnitType === 'seq_parameter_set_rbsp') {
+      track.newMetadata = true;
+      config = currentNal.config;
+      track.width = config.width;
+      track.height = config.height;
+      track.sps = [currentNal.data];
+      track.profileIdc = config.profileIdc;
+      track.levelIdc = config.levelIdc;
+      track.profileCompatibility = config.profileCompatibility;
+      h264Frame.endNalUnit();
+    } else if (currentNal.nalUnitType === 'pic_parameter_set_rbsp') {
+      track.newMetadata = true;
+      pps = currentNal.data;
+      track.pps = [currentNal.data];
+      h264Frame.endNalUnit();
+    } else if (currentNal.nalUnitType === 'access_unit_delimiter_rbsp') {
+        if (h264Frame) {
+          this.finishFrame(tags, h264Frame);
+        }
+        h264Frame = new FlvTag(FlvTag.VIDEO_TAG);
+        h264Frame.pts = currentNal.pts;
+        h264Frame.dts = currentNal.dts;
+      } else {
+        if (currentNal.nalUnitType === 'slice_layer_without_partitioning_rbsp_idr') {
+          // the current sample is a key frame
+          h264Frame.keyFrame = true;
+        }
+        h264Frame.endNalUnit();
+      }
+      h264Frame.startNalUnit();
+      h264Frame.writeBytes(currentNal.data);
+    }
+
+    this.trigger('data', {track: track, tags: tags});
+
+    // Continue with the flush process now
+    this.trigger('done');
+  };
+};
+
+VideoSegmentStream.prototype = new muxjs.Stream();
+
+/**
+ * A Stream that can combine multiple streams (ie. audio & video)
+ * into a single output segment for MSE. Also supports audio-only
+ * and video-only streams.
+ */
+CoalesceStream = function(options) {
+  // Number of Tracks per output segment
+  // If greater than 1, we combine multiple
+  // tracks into a single segment
+  this.numberOfTracks = 0;
+  this.metadataStream = options.metadataStream;
+
+  this.videoTags = [];
+  this.audioTags = [];
+  this.videoTrack = null;
+  this.audioTrack = null;
+  this.pendingCaptions = [];
+  this.pendingMetadata = [];
+  this.pendingTracks = 0;
+
+  CoalesceStream.prototype.init.call(this);
+
+  // Take output from multiple
+  this.push = function(output) {
+    // buffer incoming captions until the associated video segment
+    // finishes
+    if (output.text) {
+      return this.pendingCaptions.push(output);
+    }
+    // buffer incoming id3 tags until the final flush
+    if (output.frames) {
+      return this.pendingMetadata.push(output);
+    }
+
+    if (output.track.type === 'video') {
+      this.videoTrack = output.track;
+      this.videoTags = output.tags;
+      this.pendingTracks++;
+    }
+    if (output.track.type === 'audio') {
+      this.audioTrack = output.track;
+      this.audioTags = output.tags;
+      this.pendingTracks++;
+    }
+  };
+};
+
+CoalesceStream.prototype = new muxjs.Stream();
+CoalesceStream.prototype.flush = function() {
```

Comment:
I agree though it would involve some more refactoring. Coalesce is similar but not the same. Same with Video/AudioSegmentStream. It would be nice if more of that code were in reusable functions so that we had a finer reuse granularity than simply "stream." That should be something we should work towards when we have the time.